src/test/java/software/amazon/encryption/s3/utils/TinyBufferSubscriber.java

```diff
@@ -0,0 +1,55 @@
+package software.amazon.encryption.s3.utils;
+
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+import software.amazon.awssdk.utils.BinaryUtils;
+
+import java.nio.ByteBuffer;
+
+/**
+ * Subscriber which purposefully limits the size of buffers sent to
+ * the wrapped subscriber. This is useful for simulating adverse network conditions.
+ */
+public class TinyBufferSubscriber implements Subscriber<ByteBuffer> {
+
+    private final Subscriber<? super ByteBuffer> wrappedSubscriber;
+
+    public TinyBufferSubscriber(final Subscriber wrappedSubscriber){
+        this.wrappedSubscriber =  wrappedSubscriber;
+    }
+
+    @Override
+    public void onSubscribe(Subscription s) {
+        wrappedSubscriber.onSubscribe(s);
+    }
+
+    @Override
+    public void onNext(ByteBuffer b) {
+        int i = 0;
+        // any value below GCM block size works
+        int chunkSize = 5;
+        while (b.remaining() > chunkSize) {
+            ByteBuffer tb = b.slice();
+            tb.limit(chunkSize);
+            byte[] intermediateBuf = BinaryUtils.copyBytesFrom(tb, chunkSize);
+            b.position(i + chunkSize);
+            i += chunkSize;
+            wrappedSubscriber.onNext(ByteBuffer.wrap(intermediateBuf));
```

Comment:
nit(non-blocking): Perhaps a comment stating that splits the byte buffer into tiny chunks of byte buffers.

Reply:
I prefer to use comments for adding outside context, the code is simple enough here IMO.